LLVM build failures on Windows/Mac

<!--
Thank you for submitting a bootstrap report! Please provide detailed information to help us reproduce and diagnose the issue.
-->

### Summary

On Windows and Mac, `x.py build src/llvm-project` can fail when `download-ci-llvm = false` is set in the `config.toml`.

### Command used


Clone the rust repo, then copy the `config.example.toml` to `config.toml` and set the following:

```toml
[llvm]
download-ci-llvm = false
```

Then run:


```sh
x.py build src/llvm-project
```

### Expected behaviour

A successful build, exiting 0.

### Actual behaviour

On Mac, the result is:

```
-- Installing: /Users/ana/git/ferrocene/ferrocene/build/aarch64-apple-darwin/lld/lib/cmake/lld/LLDConfigVersion.cmake
cargo:root=/Users/ana/git/ferrocene/ferrocene/build/aarch64-apple-darwin/lld
        finished in 31.764 seconds
Building crtbegin.o and crtend.o for aarch64-apple-darwin
cargo:warning=/Users/ana/git/ferrocene/ferrocene/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:15:28: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=   15 |     __attribute__((section(".eh_frame"), aligned(sizeof(void *)))) = {};
cargo:warning=      |                            ^
cargo:warning=/Users/ana/git/ferrocene/ferrocene/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:49:24: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=   49 | __attribute__((section(".init_array"),
cargo:warning=      |                        ^
cargo:warning=/Users/ana/git/ferrocene/ferrocene/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:106:24: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=  106 | __attribute__((section(".fini_array"),
cargo:warning=      |                        ^
cargo:warning=3 errors generated.


error occurred: Command env -u IPHONEOS_DEPLOYMENT_TARGET "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx15.2" "-std=c11" "-DCRT_HAS_INITFINI_ARRAY" "-DEH_USE_FRAME_REGISTRY" "-o" "/Users/ana/git/ferrocene/ferrocene/build/aarch64-apple-darwin/native/crt/df7af2d07715af86-crtbegin.o" "-c" "/Users/ana/git/ferrocene/ferrocene/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c" with args cc did not execute successfully (status code exit status: 1).
```

On Windows, the result is:

```
-- Installing: D:/git/ferrocene/ferrocene/build/x86_64-pc-windows-msvc/llvm/include/llvm/Config/llvm-config.h
-- Installing: D:/git/ferrocene/ferrocene/build/x86_64-pc-windows-msvc/llvm/include/llvm/Support/VCSRevision.h
-- Installing: D:/git/ferrocene/ferrocene/build/x86_64-pc-windows-msvc/llvm/lib/cmake/llvm/LLVMConfigExtensions.cmake

cargo:root=D:\git\ferrocene\ferrocene\build\x86_64-pc-windows-msvc\llvm
        finished in 1242.431 seconds
Building crtbegin.o and crtend.o for x86_64-pc-windows-msvc



cargo:warning=cl : Command line warning D9002 : ignoring unknown option '-std=c11'
crtbegin.c
D:\git\ferrocene\ferrocene\src/llvm-project/compiler-rt/lib/builtins/crtbegin.c(9): fatal error C1034: stddef.h: no include path set


error occurred: Command "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.40.33807\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-O2" "-Brepro" "-std=c11" "-DCRT_HAS_INITFINI_ARRAY" "-DEH_USE_FRAME_REGISTRY" "-FoD:\\git\\ferrocene\\ferrocene\\build\\x86_64-pc-windows-msvc\\native\\crt\\89a91fa0b39682bb-crtbegin.o" "-c" "D:\\git\\ferrocene\\ferrocene\\src/llvm-project/compiler-rt/lib/builtins/crtbegin.c" with args cl.exe did not execute successfully (status code exit code: 2).


Build completed unsuccessfully in 0:39:09
```

### Bootstrap configuration (config.toml)

The same as the `config.example.toml` but:

```toml
[llvm]
download-ci-llvm = false
```

### Operating system

This is happening on, as far as I know, all Windows and Mac targets. I tested on Mac 15.2 and Windows 11 24H2.

### HEAD

bcd0683e5dce1945b5d940714742e7502883bb5c and 6a64e3b89724395d87a394ed9ebd954c474b080a both display the issue

### Additional context

After discussing with @japaric, I believe that these steps (building crtbegin.o and crtend.o) are unnecessary, bootstrap seems to claim that these steps are for cross compiling to MUSL.

We believe the issue happened sometime between #134839 (where the problem is not present) and #134987 (where the problem is now present) 


<!--
Include the complete build log in the section below.
Enable backtrace and verbose mode if possible for more detailed information e.g., with `RUST_BACKTRACE=1 ./x build -v`.
-->
(Build logs excluded as they are longer than the Github Issues limit)

[windows-log.txt](https://github.com/user-attachments/files/18480761/windows-log.txt)

[mac-log.txt](https://github.com/user-attachments/files/18480762/mac-log.txt)

@japaric asked me to repro the issue on Mac from checkout `6e1c11591fc7e98015f8e793a15c8faf0bda67f1`. Here is the output of `./x build src/llvm-project/compiler-rt/lib/crt `:

```
Submodule path 'src/llvm-project': checked out '104d0d16c3c7c3fef2435fef6efb2d57b70fff73'
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-aarch64-apple-darwin.tar.xz
################################################################################################################################################################# 100.0%
extracting /Users/ana/git/rust-lang/rust/build/cache/2024-10-16/rustfmt-nightly-aarch64-apple-darwin.tar.xz to /Users/ana/git/rust-lang/rust/build/aarch64-apple-darwin/rustfmt
downloading https://static.rust-lang.org/dist/2024-10-16/rustc-nightly-aarch64-apple-darwin.tar.xz
################################################################################################################################################################# 100.0%
extracting /Users/ana/git/rust-lang/rust/build/cache/2024-10-16/rustc-nightly-aarch64-apple-darwin.tar.xz to /Users/ana/git/rust-lang/rust/build/aarch64-apple-darwin/rustfmt
Building crtbegin.o and crtend.o for aarch64-apple-darwin
cargo:warning=/Users/ana/git/rust-lang/rust/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:15:28: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=   15 |     __attribute__((section(".eh_frame"), aligned(sizeof(void *)))) = {};
cargo:warning=      |                            ^
cargo:warning=/Users/ana/git/rust-lang/rust/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:49:24: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=   49 | __attribute__((section(".init_array"),
cargo:warning=      |                        ^
cargo:warning=/Users/ana/git/rust-lang/rust/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c:106:24: error: argument to 'section' attribute is not valid for this target: mach-o section specifier requires a segment and section separated by a comma
cargo:warning=  106 | __attribute__((section(".fini_array"),
cargo:warning=      |                        ^
cargo:warning=3 errors generated.


error occurred: Command env -u IPHONEOS_DEPLOYMENT_TARGET "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-darwin" "-mmacosx-version-min=15.2" "-std=c11" "-DCRT_HAS_INITFINI_ARRAY" "-DEH_USE_FRAME_REGISTRY" "-o" "/Users/ana/git/rust-lang/rust/build/aarch64-apple-darwin/native/crt/201e8d6b412081ae-crtbegin.o" "-c" "/Users/ana/git/rust-lang/rust/src/llvm-project/compiler-rt/lib/builtins/crtbegin.c" with args cc did not execute successfully (status code exit status: 1).


Build completed unsuccessfully in 0:01:13
Took 1min 14sec 350ms
```

--- a/src/bootstrap/src/core/build_steps/compile.rs
+++ b/src/bootstrap/src/core/build_steps/compile.rs
@@ -339,7 +339,7 @@ fn copy_self_contained_objects(
     // to using gcc from a glibc-targeting toolchain for linking.
     // To do that we have to distribute musl startup objects as a part of Rust toolchain
     // and link with them manually in the self-contained mode.
-    if target.contains("musl") && !target.contains("unikraft") {
+    if target.needs_crt_begin_end() {
         let srcdir = builder.musl_libdir(target).unwrap_or_else(|| {
             panic!("Target {:?} does not have a \"musl-libdir\" key", target.triple)
         });

--- a/src/bootstrap/src/core/build_steps/llvm.rs
+++ b/src/bootstrap/src/core/build_steps/llvm.rs
@@ -1295,7 +1295,9 @@ impl Step for CrtBeginEnd {
     }
 
     fn make_run(run: RunConfig<'_>) {
-        run.builder.ensure(CrtBeginEnd { target: run.target });
+        if run.target.needs_crt_begin_end() {
+            run.builder.ensure(CrtBeginEnd { target: run.target });
+        }
     }
 
     /// Build crtbegin.o/crtend.o for musl target.

--- a/src/bootstrap/src/core/config/config.rs
+++ b/src/bootstrap/src/core/config/config.rs
@@ -573,6 +573,10 @@ impl TargetSelection {
         env::var("OSTYPE").is_ok_and(|v| v.to_lowercase().contains("cygwin"))
     }
 
+    pub fn needs_crt_begin_end(&self) -> bool {
+        self.contains("musl") && !self.contains("unikraft")
+    }
+
     /// Path to the file defining the custom target, if any.
     pub fn filepath(&self) -> Option<&Path> {
         self.file.as_ref().map(Path::new)

